src/esnext.coffee

```diff
@@ -0,0 +1,154 @@
+    shasum.update('"', 'utf8')
+  else if typeof value in ['boolean', 'number']
+    shasum.update(value.toString(), 'utf8')
+  else if typeof value is null
+    shasum.update('null', 'utf8')
+  else if Array.isArray value
+    shasum.update('[', 'utf8')
+    for item in value
+      updateDigestForJsonValue(shasum, item)
+      shasum.update(',', 'utf8')
+    shasum.update(']', 'utf8')
+  else
+    # value must be an object: be sure to sort the keys.
+    keys = Object.keys value
+    keys.sort()
+
+    shasum.update('{', 'utf8')
+    for key in keys
+      updateDigestForJsonValue(shasum, key)
+      shasum.update(': ', 'utf8')
+      updateDigestForJsonValue(shasum, value[key])
+      shasum.update(',', 'utf8')
+    shasum.update('}', 'utf8')
+
+create6to5VersionAndOptionsDigest = (version, options) ->
+  shasum = crypto.createHash('sha1')
+  # Include the version of 6to5 in the hash.
+  shasum.update('6to5-core', 'utf8')
+  shasum.update('\0', 'utf8')
+  shasum.update(version, 'utf8')
+  shasum.update('\0', 'utf8')
+  updateDigestForJsonValue(shasum, options)
+  shasum.digest('hex')
+
+cacheDir = path.join(fs.absolute('~/.atom'), 'compile-cache')
+jsCacheDir = path.join(
+    cacheDir,
+    create6to5VersionAndOptionsDigest(to5.version, defaultOptions),
+    'js')
+
+getCachePath = (sourceCode) ->
+  digest = crypto.createHash('sha1').update(sourceCode, 'utf8').digest('hex')
+  path.join(jsCacheDir, "#{digest}.js")
+
+getCachedJavaScript = (cachePath) ->
+  if fs.isFileSync(cachePath)
+    try
+      cachedJavaScript = fs.readFileSync(cachePath, 'utf8')
+      stats.hits++
+      return cachedJavaScript
+  null
+
+# Returns the 6to5 options that should be used to transpile filePath.
+createOptions = (filePath) ->
+  options = filename: filePath
+  for key, value of defaultOptions
+    options[key] = value
+  options
+
+# Function that obeys the contract of an entry in the require.extensions map.
+# Returns the transpiled version of the JavaScript code at filePath, which is
+# either generated on the fly or pulled from cache.
+loadFile = (module, filePath) ->
+  sourceCode = fs.readFileSync(filePath, 'utf8')
+  unless sourceCode.startsWith('"use 6to5"') or sourceCode.startsWith("'use 6to5'")
+    module._compile(sourceCode, filePath)
+    return
+
+  cachePath = getCachePath(sourceCode)
+  js = getCachedJavaScript(cachePath)
+
+  unless js
+    options = createOptions filePath
+    try
+      js = to5.transform(sourceCode, options).code
+      stats.misses++
+    catch error
+      console.error('Error compiling %s: %o', filePath, error)
+      throw error
+
+    try
+      fs.writeFileSync(cachePath, js)
+    catch error
+      console.error('Error writing to cache at %s: %o', cachePath, error)
+      throw error
+
+  module._compile(js, filePath)
+
+register = () ->
+  require.extensions['.js'] = loadFile
+
+module.exports =
+  register: register
+  getCacheMisses: -> stats.misses
+  getCacheHits: -> stats.hits
```

Comment:
Down the road, if you want to see these in Timecop, they can be added [here](https://github.com/atom/timecop/blob/12df77d91c7686e65d9825ffdb4db506c70b4b25/lib/cache-panel-view.coffee#L26-L29)